command/monitor/monitor.go

```diff
@@ -0,0 +1,150 @@
+package monitor
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	log "github.com/hashicorp/go-hclog"
+)
+
+// Monitor provides a mechanism to stream logs using go-hclog
+// InterceptLogger and SinkAdapter. It allows streaming of logs
+// at a different log level than what is set on the logger.
+type Monitor interface {
+	// Start returns a channel of log messages which are sent
+	// every time a log message occurs
+	Start() <-chan []byte
+
+	// Stop de-registers the sink from the InterceptLogger
+	// and closes the log channels
+	Stop()
+}
+
+// monitor implements the Monitor interface
+type monitor struct {
+	sink log.SinkAdapter
+
+	// logger is the logger we will be monitoring
+	logger log.InterceptLogger
+
+	// logCh is a buffered chan where we send logs when streaming
+	logCh chan []byte
+
+	// doneCh coordinates the shutdown of logCh
+	doneCh chan struct{}
+
+	// droppedCount is the current count of messages
+	// that were dropped from the logCh buffer.
+	droppedCount uint32
+	bufSize      int
+	// dropCheckInterval is the amount of time we should
+	// wait to check for dropped messages. Defaults
+	// to 3 seconds
+	dropCheckInterval time.Duration
+}
+
+// NewMonitor creates a new Monitor. Start must be called in order to actually start
+// streaming logs. buf is the buffer size of the channel that sends log messages.
+func NewMonitor(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) (Monitor, error) {
+	return newMonitor(buf, logger, opts)
+}
+
+func newMonitor(buf int, logger log.InterceptLogger, opts *log.LoggerOptions) (*monitor, error) {
+	if buf <= 0 {
+		return nil, fmt.Errorf("buf must be greater than zero")
+	}
+
+	sw := &monitor{
+		logger:            logger,
+		logCh:             make(chan []byte, buf),
+		doneCh:            make(chan struct{}),
+		bufSize:           buf,
+		dropCheckInterval: 3 * time.Second,
+	}
+
+	opts.Output = sw
+	sink := log.NewSinkAdapter(opts)
+	sw.sink = sink
+
+	return sw, nil
+}
+
+// Stop deregisters the sink and stops the monitoring process
+func (d *monitor) Stop() {
+	d.logger.DeregisterSink(d.sink)
+	close(d.doneCh)
+}
+
+// Start registers a sink on the monitor's logger and starts sending
+// received log messages over the returned channel.
+func (d *monitor) Start() <-chan []byte {
```

Comment:
After stop is called this function will likely panic trying to access d.doneCh since it will be closed. You should probably save a state var indicating if Stop has been called and refuse to Start again. Or, re-initialize d.doneCh when Start is called. Technically you never close logCh; Write probably should when it terminates, but in that case when Start is called, if you want to allow re-starting, you'd need to reinitialize that.

It'd be good to note explicitly that this struct is not threadsafe.

Reply:
Actually rather than indicating if Stop has been called, use a check and set atomic at the beginning of Start to indicate if Start has been called and disallow running it again.

Reply:
Will do

Reply:
Thinking on your comment a little more, I'm confused. The only time we try to access `d.doneCh` in this function is to attempt to read from it. I thought reading from a closed channel would just return immediately, not panic. I feel like I'm missing your point.

I'll add the check and set atomic at the beginning of `Start`, regardless.